inc/drivers/MicroBitFileSystem.h

```diff
@@ -0,0 +1,571 @@
+#ifndef MICROBIT_FILE_SYSTEM_H
+#define MICROBIT_FILE_SYSTEM_H
+#include "MicroBitConfig.h"
+#include "MicroBitFlash.h"
+#include <stdint.h>
+
+/**
+  * @file MicroBitFileSystem.h
+  * @author Alex King
+  * @date 6 Mar 2016
+  *
+  * @brief API for POSIX-like file system interface: open/close/read/write.
+  *
+  * -- Basic overview of how the file system works:
+  * All file meta data is stored in the first page. Each file has a record
+  * entry (struct FileTableEntry_t), which stores:
+  *  - Filename
+  *  - File size
+  *  - An ordered list of pages/blocks that store the file data itself.
+  *
+  * The table housing these records is referred to as the File Table (FT)
+  * and looks something like this:
+  *  {"file1.txt",   5000,   {1,2,3,6}}  // 5000-byte file, in blocks 1,2,3,6
+  *  {"file02.txt",   200,   {5}}        // 200 byte file in a single block
+  *
+  * The first FT entry is reserved, and is used only to store a list of
+  * currently available blocks for writing.
+  * Thus to expand a file size, a block must be removed from this list, and
+  * added to that files blocklist.
+  *
+  * Since there can only be one FT, the number of entries is restricted
+  * to how many can fit in a single flash page. Hence, this is the
+  * maximum number of files that can be stored.
+  *
+  * The filesize is stored in the FT, but cached in the mb_fd struct.
+  * Writes to the FT filesize are made sparingly in write() and close(), to
+  * reduce erasures.
+  * @warning close() *must* be called to ensure the FT is synched, before
+  * the microbit is finished.
+  *
+  * -- API overview.
+  * Source code is divided logically into two parts:
+  * ft_*() functions - to access and modify the FT entries.
+  *                    E.g., to create a new file, change a files size,
+  *                    or append new blocks to the file.
+  *                    These are private methods internal to the class.
+  *
+  * read/write/seek/remove - POSIX-style file access functions.
+  *                          These are the methods used to interact with the
+  *                          file system.
+  *
+  * Example:
+  * @code
+  * MicroBitFileSystem f;
+  * int fd = f.open("myFile.txt", MB_WRITE|MB_CREAT);
+  * if(fd < 0) {
+  *   print("couldn't open a new file");
+  *   exit;
+  * }
+  * if(f.write(fd, "Hello, World!\n",14) != 14) {
+  *   print("error writing!")
+  *   exit
+  * }
+  * f.close(fd);
+  * @endcode
+  *
+  * -- Notes
+  * - There should exist only one instance of this class.
+  * @todo implement simple wear levelling for file data writes.
+  * @todo implement MB_APPEND flag in open()
+  */
+
+// Configuration options.
+
+#define PAGE_SIZE 1024  // Size of page in flash.
```

Comment:
Does this need to match the erase-size of a page?  Is this the minimum amount of bytes that a given file must take up (eg a file with 1 character takes up 1024 bytes)?